pkg/filter/filter.go

```diff
@@ -0,0 +1,29 @@
+package filter
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-admission-controller/pkg/release"
+)
+
+// IsObjectReconciledByLegacyRelease checks if the object is reconciled by an operator which is the
+// part of a legacy Giant Swarm release (a release that does not have Cluster API controllers).
+func IsObjectReconciledByLegacyRelease(ctx context.Context, ctrlCache client.Reader, objectMeta metav1.Object) (bool, error) {
```

Comment:
@tuommaki the filtering func in now just calling two new funcs. Good suggestion to split this, it's easier to understand it this way.

Reply:
The code is indeed more readable, but I think the original problem of two clients was slightly missed here. Now there's only one client used here and it's the caching one. This has the effect that now `Cluster` is also cached in the informer. It might not be a bad idea on itself, but should be conscious decision.

I'm thinking that these problems with two different clients and challenges in wiring / carrying them through call chains are a bit of a smell of non-optimal architecture. The functions we have here are theoretically just something that should work on plain data. One should be able to implement them in pure functions given that the input data was fetched beforehand.

The API of function signatures is as generic as it can be while somewhat typed, but the implementation itself is quite specific since it involves knowledge about `Cluster` type and how to find it.

Could the design be such that all necessary objects are fetched on top level and then passed as data to these functions? That way we could better control the API calls and avoid accidental re-fetching in different execution paths while making functions signatures more clear on intent and allowing easier testing of them.

Example:
Top level function to fetch the release. Note that `Cluster` was already fetched and the passed reader is only used to fetch the `Release` object itself.
`func FindRelease(ctx context.Context, ctrlCache client.Reader, cluster capi.Cluster, objectMeta metav1.Object) (releasev1alpha1.Release, error)`

The logic that knows whether it's legacy release or not. No other objects nor clients needed since the `Release` itself is here:
`func IsLegacyRelease(release releasev1alpha1.Release) (bool, error)`

I think it could be useful if we paired on this a bit so that there's not that long latency through asynchronous reviews?

Reply:
> This has the effect that now Cluster is also cached in the informer. It might not be a bad idea on itself, but should be conscious decision.

Oh, I was not aware of that.

> One should be able to implement them in pure functions given that the input data was fetched beforehand.
> ...
> Could the design be such that all necessary objects are fetched on top level and then passed as data to these functions? That way we could better control the API calls and avoid accidental re-fetching in different execution paths while making functions signatures more clear on intent and allowing easier testing of them.

I see, yeah that would be probably better, to basically make the function a stateless "utility" function, and make consumers (that know about cached vs not cached) talk to the API.